source/common/grpc/google_async_client_impl.cc

```diff
@@ -144,25 +192,50 @@ void GoogleAsyncStreamImpl::writeQueued() {
 
   if (!msg.buf_.valid()) {
     ASSERT(msg.end_stream_);
-    rw_->WritesDone(tag(Operation::WriteLast));
+    rw_->WritesDone(&write_last_tag_);
+    ++inflight_tags_;
   } else if (msg.end_stream_) {
     grpc::WriteOptions write_options;
-    rw_->WriteLast(msg.buf_.value(), write_options, tag(Operation::WriteLast));
+    rw_->WriteLast(msg.buf_.value(), write_options, &write_last_tag_);
+    ++inflight_tags_;
   } else {
-    rw_->Write(msg.buf_.value(), tag(Operation::Write));
+    rw_->Write(msg.buf_.value(), &write_tag_);
+    ++inflight_tags_;
   }
+  ENVOY_LOG(trace, "Write op dispatched");
 }
 
-void GoogleAsyncStreamImpl::handleOpCompletion(Operation op, bool ok) {
-  ENVOY_LOG(trace, "handleOpCompletion {} {}", op, ok);
-  // Ignore op completions while CQ is shutting down.
-  if (cq_shutdown_in_progress_) {
+void GoogleAsyncStreamImpl::onCompletedOp() {
+  std::unique_lock<std::mutex> lock(completed_ops_lock_);
```

Comment:
Do you want to drop the lock when calling handleOpCompletion()? Or does it not matter that much for contention?

Reply:
Yeah, good point, should release it.